### PR TITLE
 feat: Implement futures-0.2 support

### PR DESCRIPTION
**Pull request by Marwes**
https://github.com/tokio-rs/tokio-rfcs/pull/2

- [x] Re-implement vectored read and write

BREAKING CHANGE

The `needs_*` methods has been removed as `PollEvented` do not have them anymore
The crate now uses the tokio-reactor crate instead of tokio_core
Implementation of the (depreceated) `tokio_core::io::Io` trait has been removed.

### CODE REVIEW

**Comment by carllerche** [comment]
Thanks for doing this.

I would like to hold off on this for a little bit. Ideally, these libs would not depend on `tokio` directly. I plan on splitting out the reactor into a separate crate, and then `tokio-uds` can be updated to depend on *that* crate.

I should be able to get that work shipped pretty quickly (ideally this week).

**Comment by Marwes** [comment]
No, rush at all. Just one of these changes that are just as easy to send a PR for as they are to open an issue.

**Comment by ufobat** [comment]
Since the last update on tokio, it would be awesome if this would work with the current tokio version.

**Comment by Marwes** [comment]
@ufobat Partially moved the code over to use tokio and futures 0.2-alpha. There are parts of `tokio` that is not yet updated to work with futures 0.2 yet however so can't complete it yet.

**Comment by ufobat** [comment]
thank you!

**Comment by carllerche** [comment]
Thanks for updating the PR! Let me take a look at this now.

**Comment by carllerche** [comment]
@Marwes Would it be possible to add an `unstable-futures` feature flag here as well?

Basically, follow the pattern that `tokio-tcp` takes? https://github.com/tokio-rs/tokio/tree/master/tokio-tcp

**Comment by Marwes** [comment]
> @Marwes Would it be possible to add an unstable-futures feature flag here as well?

Sure. I will look into it tomorrow.

**Comment by Marwes** [comment]
@carllerche It should be good now I think but I may have accidentally messed up something as I were restoring the old impls.

**Comment by Marwes** [comment]
ping @carllerche

**Comment by carllerche** [comment]
@Marwes Thanks for doing this. Sorry for the delay. I did a quick pass but I will have to focus on it more tomorrow. I will try to leave some quick inline notes.

The first thing that stands out is that there are a number of unrelated changes that make the PR harder to review. I will try to point them out inline. Would it be possible to revert those for this PR?

File: Cargo.toml [hunk 1]
```diff
@@ -1,27 +1,33 @@
...
-iovec = "0.1"
-libc = "0.2"
-log = "0.4"
-mio = "0.6.5"
-mio-uds = "0.6.4"
-tokio-core = "0.1"
-tokio-io = "0.1"
+[package]
+name = "tokio-uds"
+version = "0.1.7"
```
Comment (carllerche) [inline Cargo.toml]: This will be a breaking change, so this should be bumped to v0.2.0.

File: src/frame.rs [hunk 2]
```diff
@@ -2,7 +2,12 @@ use std::io;
 use std::os::unix::net::SocketAddr;
 use std::path::PathBuf;
 
-use futures::{Async, Poll, Stream, Sink, StartSend, AsyncSink};
+use futures::{Async, AsyncSink, Poll, Sink, StartSend, Stream};
```
Comment (carllerche) [inline src/frame.rs]: This does not seem related to the PR?

File: src/frame.rs [hunk 3]
```diff
@@ -50,8 +55,7 @@ pub trait UnixDatagramCodec {
     ///
     /// The encode method also determines the destination to which the buffer
     /// should be directed, which will be returned as a `SocketAddr`.
-    fn encode(&mut self, msg: Self::Out, buf: &mut Vec<u8>)
-              -> io::Result<PathBuf>;
+    fn encode(&mut self, msg: Self::Out, buf: &mut Vec<u8>) -> io::Result<PathBuf>;
```
Comment (carllerche) [inline src/frame.rs]: This does not seem related to the PR?

File: src/frame.rs [hunk 4]
```diff
@@ -101,19 +119,21 @@ impl<C: UnixDatagramCodec> Sink for UnixDatagramFramed<C> {
         trace!("flushing framed transport");
 
         if self.wr.is_empty() {
-            return Ok(Async::Ready(()))
+            return Ok(Async::Ready(()));
```
Comment (carllerche) [inline src/frame.rs]: This does not seem related to the PR?

File: src/frame.rs [hunk 5]
```diff
@@ -101,19 +119,21 @@ impl<C: UnixDatagramCodec> Sink for UnixDatagramFramed<C> {
...
+        let n = try_ready!(self.socket.send_to(&self.wr, &self.out_addr));
         trace!("written {}", n);
         let wrote_all = n == self.wr.len();
         self.wr.clear();
         if wrote_all {
             Ok(Async::Ready(()))
         } else {
-            Err(io::Error::new(io::ErrorKind::Other,
-                               "failed to write entire datagram to socket"))
+            Err(io::Error::new(
```
Comment (carllerche) [inline src/frame.rs]: This does not seem related to the PR?

File: src/lib.rs [hunk 6]
```diff
@@ -33,33 +39,84 @@ use std::os::unix::prelude::*;
 use std::path::Path;
 
 use bytes::{Buf, BufMut};
-use futures::{Future, Poll, Async, Stream};
-use futures::sync::oneshot;
+use futures::{Async, Future, Poll, Stream};
```
Comment (carllerche) [inline src/lib.rs]: This does not seem related to the PR?

File: src/lib.rs [hunk 7]
```diff
@@ -33,33 +39,84 @@ use std::os::unix::prelude::*;
...
+            Ok($crate::futures2::Async::Ready(t)) => t,
+            Ok($crate::futures2::Async::Pending) => return Ok($crate::futures2::Async::Pending),
+            Err(e) => return Err(From::from(e)),
+        }
+    };
+}
 
 mod frame;
-pub use frame::{UnixDatagramFramed, UnixDatagramCodec};
+pub use frame::{UnixDatagramCodec, UnixDatagramFramed};
```
Comment (carllerche) [inline src/lib.rs]: This does not seem related to the PR?

File: src/lib.rs [hunk 8]
```diff
@@ -69,23 +126,21 @@ impl UnixListener {
     ///
     /// The returned listener will be associated with the given event loop
     /// specified by `handle` and is ready to perform I/O.
-    pub fn from_listener(listener: net::UnixListener, handle: &Handle)
-                         -> io::Result<UnixListener> {
+    pub fn from_listener(listener: net::UnixListener, handle: Handle) -> io::Result<UnixListener> {
```
Comment (carllerche) [inline src/lib.rs]: This does not seem related to the PR? Also, this probably should be `&Handle`.

Also, could this be named `from_std`?

File: src/lib.rs [hunk 9]
```diff
@@ -33,33 +39,84 @@ use std::os::unix::prelude::*;
...
-        where P: AsRef<Path>
+    pub fn bind<P>(path: P) -> io::Result<UnixListener>
+    where
+        P: AsRef<Path>,
+    {
+        UnixListener::bind_handle(path.as_ref(), Handle::default())
+    }
+
+    /// Creates a new `UnixListener` bound to the specified path.
+    pub fn bind_handle<P>(path: P, handle: Handle) -> io::Result<UnixListener>
```
Comment (carllerche) [inline src/lib.rs]: This function can be removed.

File: src/lib.rs [hunk 10]
```diff
@@ -238,8 +310,7 @@ impl UnixStream {
     ///
     /// The returned stream will be associated with the given event loop
     /// specified by `handle` and is ready to perform I/O.
-    pub fn from_stream(stream: net::UnixStream, handle: &Handle)
-                       -> io::Result<UnixStream> {
+    pub fn from_stream(stream: net::UnixStream, handle: &Handle) -> io::Result<UnixStream> {
```
Comment (carllerche) [inline src/lib.rs]: Could this be renamed `from_std`?

File: src/lib.rs [hunk 11]
```diff
@@ -220,10 +279,23 @@ impl UnixStream {
...
+    }
+
+    /// Connects to the socket named by `path`.
+    ///
+    /// This function will create a new unix socket and connect to the path
+    /// specified, associating the returned stream with the provided
     /// event loop's handle.
-    pub fn connect<P>(p: P, handle: &Handle) -> io::Result<UnixStream>
-        where P: AsRef<Path>
+    pub fn connect_handle<P>(p: P, handle: &Handle) -> io::Result<UnixStream>
```
Comment (carllerche) [inline src/lib.rs]: This fn can be removed.

File: src/lib.rs [hunk 12]
```diff
@@ -419,18 +593,32 @@ unsafe fn read_ready<B: BufMut>(buf: &mut B, raw_fd: RawFd) -> isize {
     let b15: &mut [u8] = &mut [0];
     let b16: &mut [u8] = &mut [0];
     let mut bufs: [&mut IoVec; 16] = [
-        b1.into(), b2.into(), b3.into(), b4.into(),
-        b5.into(), b6.into(), b7.into(), b8.into(),
-        b9.into(), b10.into(), b11.into(), b12.into(),
-        b13.into(), b14.into(), b15.into(), b16.into(),
+        b1.into(),
```
Comment (carllerche) [inline src/lib.rs]: This does not seem related to the PR?

File: src/lib.rs [hunk 13]
```diff
@@ -468,18 +656,18 @@ unsafe fn write_ready<B: Buf>(buf: &mut B, raw_fd: RawFd) -> isize {
     static DUMMY: &[u8] = &[0];
     let iovec = <&IoVec>::from(DUMMY);
     let mut bufs = [
-        iovec, iovec, iovec, iovec,
-        iovec, iovec, iovec, iovec,
-        iovec, iovec, iovec, iovec,
-        iovec, iovec, iovec, iovec,
+        iovec, iovec, iovec, iovec, iovec, iovec, iovec, iovec, iovec, iovec, iovec, iovec, iovec,
```
Comment (carllerche) [inline src/lib.rs]: This does not seem related to the PR?

File: src/lib.rs [hunk 14]
```diff
@@ -530,7 +718,8 @@ pub struct UnixDatagram {
 impl UnixDatagram {
     /// Creates a new `UnixDatagram` bound to the specified path.
     pub fn bind<P>(path: P, handle: &Handle) -> io::Result<UnixDatagram>
-        where P: AsRef<Path>
+    where
```
Comment (carllerche) [inline src/lib.rs]: This does not seem related to the PR?

File: src/lib.rs [hunk 15]
```diff
@@ -557,15 +746,13 @@ impl UnixDatagram {
     ///
     /// The returned datagram will be associated with the given event loop
     /// specified by `handle` and is ready to perform I/O.
-    pub fn from_datagram(datagram: net::UnixDatagram, handle: &Handle)
-                       -> io::Result<UnixDatagram> {
+    pub fn from_datagram(datagram: net::UnixDatagram, handle: &Handle) -> io::Result<UnixDatagram> {
```
Comment (carllerche) [inline src/lib.rs]: This does not seem related to the PR?

File: src/lib.rs [hunk 16]
```diff
@@ -557,15 +746,13 @@ impl UnixDatagram {
...
         let s = try!(mio_uds::UnixDatagram::from_datagram(datagram));
         UnixDatagram::new(s, handle)
     }
 
-    fn new(socket: mio_uds::UnixDatagram, handle: &Handle)
-           -> io::Result<UnixDatagram> {
-        let io = try!(PollEvented::new(socket, handle));
+    fn new(socket: mio_uds::UnixDatagram, handle: &Handle) -> io::Result<UnixDatagram> {
```
Comment (carllerche) [inline src/lib.rs]: This (first line) does not seem related to the PR?

File: src/lib.rs [hunk 17]
```diff
@@ -631,34 +796,72 @@ impl UnixDatagram {
...
+        if is_wouldblock(&r) {
+            self.io.clear_read_ready2(cx, Ready::readable())?;
+        }
+        r.map(Async::Ready)
     }
 
     /// Returns a future for receiving a datagram. See the documentation on RecvDgram for details.
     pub fn recv_dgram<T>(self, buf: T) -> RecvDgram<T>
-        where T: AsMut<[u8]>
+    where
```
Comment (carllerche) [inline src/lib.rs]: This does not seem related to the PR?

File: src/lib.rs [hunk 18]
```diff
@@ -690,22 +894,65 @@ impl UnixDatagram {
     /// will return an error if the socket has not already been connected.
     ///
     /// On success, returns the number of bytes written.
-    pub fn send(&self, buf: &[u8]) -> io::Result<usize> {
-        if self.io.poll_write().is_not_ready() {
-            return Err(would_block())
+    pub fn send(&mut self, buf: &[u8]) -> Poll<usize, io::Error> {
```
Comment (carllerche) [inline src/lib.rs]: Is adding `&mut` here required?

File: src/lib.rs [hunk 19]
```diff
@@ -690,22 +894,65 @@ impl UnixDatagram {
...
+        let r = self.io.get_ref().send(buf);
+        if is_wouldblock(&r) {
+            self.io.clear_write_ready2(cx)?;
+        }
+        r.map(Async::Ready)
+    }
 
     /// Returns a future sending the data in buf to the socket at path.
     pub fn send_dgram<T, P>(self, buf: T, path: P) -> SendDgram<T, P>
-        where T: AsRef<[u8]>,
```
Comment (carllerche) [inline src/lib.rs]: This does not seem related to the PR?

File: src/lib.rs [hunk 20]
```diff
@@ -750,7 +997,8 @@ impl UnixDatagram {
     /// which will break them into separate objects, allowing them to interact
     /// more easily.
     pub fn framed<C>(self, codec: C) -> UnixDatagramFramed<C>
-        where C: UnixDatagramCodec,
+    where
```
Comment (carllerche) [inline src/lib.rs]: This does not seem related to the PR?

File: src/lib.rs [hunk 21]
```diff
@@ -795,26 +1043,74 @@ enum SendDgramState<T, P> {
 }
 
 impl<T, P> Future for SendDgram<T, P>
-    where T: AsRef<[u8]>,
-          P: AsRef<Path>
+where
```
Comment (carllerche) [inline src/lib.rs]: This does not seem related to the PR?

File: src/ucred.rs [hunk 22]
```diff
@@ -1,4 +1,4 @@
-use libc::{uid_t, gid_t};
+use libc::{gid_t, uid_t};
```
Comment (carllerche) [inline src/ucred.rs]: This does not seem related to the PR?

File: src/ucred.rs [hunk 23]
```diff
@@ -17,7 +17,7 @@ pub use self::impl_macos::get_peer_cred;
 
 #[cfg(any(target_os = "linux", target_os = "android"))]
 pub mod impl_linux {
-    use libc::{getsockopt, SOL_SOCKET, SO_PEERCRED, c_void, socklen_t};
+    use libc::{c_void, getsockopt, socklen_t, SOL_SOCKET, SO_PEERCRED};
```
Comment (carllerche) [inline src/ucred.rs]: This does not seem related to the PR?

File: src/ucred.rs [hunk 24]
```diff
@@ -28,7 +28,11 @@ pub mod impl_linux {
         unsafe {
             let raw_fd = sock.as_raw_fd();
 
-            let mut ucred = ucred { pid: 0, uid: 0, gid: 0 };
+            let mut ucred = ucred {
```
Comment (carllerche) [inline src/ucred.rs]: This does not seem related to the PR?

**Comment by Marwes** [comment]
@carllerche Sorry about the rustfmting, I broke it out to a separate PR #32 . Diff without that commit at https://github.com/tokio-rs/tokio-uds/pull/29/files/c6826ec86c5b6f6f831516ba044521ac52190900..14e9c084691cd6ece741ecc8d0adc676a23b3c3e .

File: src/lib.rs (hunk 9 above)
Comment (Marwes) [reply src/lib.rs]: I was under the impression that there should both be a method that takes an explicit handle and one without (that grabs the default handle). Is that not the case anymore?

File: src/lib.rs (hunk 8 above)
Comment (Marwes) [reply src/lib.rs]: > This does not seem related to the PR? Also, this probably should be &Handle

A `Handle` is needed to construct the `PollEvented` that wraps the mio stream so I had to take ownership and store a handle in the listener for that. Figured it were better to be explicit about the owning part.

**Comment by arcnmx** [comment]
The `tokio-reactor = { features=["unstable-features"] }` bit seems wrong, it should be changed to `unstable-futures = ["futures-core", "futures-io", "futures-sink", "tokio-reactor/unstable-features"]`?

Also the `futures-*` deps can probably be updated to final 0.2.0 versions, which are now released.

**Comment by Marwes** [comment]
Broke out the tokio 0.1 support into #36.

**Comment by carllerche** [comment]
I'm going to close this as most of the important work has been landed via other PRs. Thanks again!
